### ISSUE TITLE
TypeError: Cannot read property 'id' of null

### ISSUE THREAD

**Issue by nealeb4zod** [rfennell/AzurePipelines#1171]
### Azure DevOps Extensions

Generate Release Notes (Node Cross Platform)

### Platform

Azure DevOps Services

### Describe the bug

Hi,

We are getting the following error when generating release notes through our build pipeline:
```
2021-12-16T11:06:14.9289668Z Looking for parents of WI [972]
2021-12-16T11:06:14.9292141Z The most common runtime reason for the task to fail is due API ECONNRESET issues. To avoid this failing the pipeline these will be treated as warnings and an attempt to generate any release notes possible
2021-12-16T11:06:14.9307333Z TypeError: Cannot read property 'id' of null
2021-12-16T11:06:14.9307615Z Total Builds: [0]
2021-12-16T11:06:14.9308052Z Total Commits: [196]
2021-12-16T11:06:14.9308455Z Total Workitems: [58]
2021-12-16T11:06:14.9308859Z Total Related Workitems (Parent/Children): [74]
2021-12-16T11:06:14.9309246Z Total Release Tests: [0]
2021-12-16T11:06:14.9309603Z Total Tests: [0]
2021-12-16T11:06:14.9309947Z Total Manual Test Runs: [0]
2021-12-16T11:06:14.9310412Z Total Manual Test Configurations: [0]
2021-12-16T11:06:14.9310822Z Total Pull Requests: [0]
2021-12-16T11:06:14.9311192Z Total Indirect Pull Requests: [0]
2021-12-16T11:06:14.9311590Z Total Consumed Artifacts: [1]
2021-12-16T11:06:14.9311954Z Total WIQL Workitems: [0]
2021-12-16T11:06:14.9312994Z     at allRelatedWorkItems.find.element (D:\a\_tasks\XplatGenerateReleaseNotes_7b0a5887-75c4-4ade-a915-f92a93c934cb\3.76.4\ReleaseNotesFunctions.js:777:74)
2021-12-16T11:06:14.9313800Z     at Array.find (<anonymous>)
2021-12-16T11:06:14.9314546Z     at D:\a\_tasks\XplatGenerateReleaseNotes_7b0a5887-75c4-4ade-a915-f92a93c934cb\3.76.4\ReleaseNotesFunctions.js:777:50
2021-12-16T11:06:14.9315272Z     at Generator.next (<anonymous>)
2021-12-16T11:06:14.9316033Z     at fulfilled (D:\a\_tasks\XplatGenerateReleaseNotes_7b0a5887-75c4-4ade-a915-f92a93c934cb\3.76.4\ReleaseNotesFunctions.js:5:58)
2021-12-16T11:06:14.9316840Z     at process._tickCallback (internal/process/next_tick.js:68:7)
```
We've previously seen the same error with at least one other WI, however I can't see any difference between them and WIs that do not cause the error.

Running the same build though the test console does not reproduce the error unfortunately - the WI is skipped as it's parent is already in the list.

Build inputs:
```
inputs:
                templateLocation: 'File'
                templatefile: '$(Build.SourcesDirectory)/ReleaseNotes/release-notes.html'
                customHandlebarsExtensionFile: 'handlebarExtensions.js'
                customHandlebarsExtensionFolder: '$(Build.SourcesDirectory)/ReleaseNotes'
                outputfile: '$(Build.SourcesDirectory)/ReleaseNotes/_WhatsNew_${{ parameters.buildConfiguration }}.vbhtml'
                sortCS: true
                sortWi: true
                dumpPayloadToConsole: false
                dumpPayloadToFile: false
                replaceFile: false
                appendToFile: true
                getParentsAndChildren: true
                getAllParents: true
                getIndirectPullRequests: False
                stopOnError: False
                considerPartiallySuccessfulReleases: False
```

Any pointers would be greatly appreciated.

### Repo Steps

Generate release notes through a build pipeline with the above inputs.

### Logging Information

[releasenoteslogs.txt](https://github.com/rfennell/AzurePipelines/files/7727035/releasenoteslogs.txt)

**Comment by rfennell** [comment]
There a few things going on here....

The fault in the `dumpJsonPayload` method, is it non-fatal i.e. is it generating the release notes or an empty file?

The failing line, given you are not dumping the payload, must the `let data = JSON.stringify(payload);`, so we have a bad payload, as we could expect from the initial `Cannot read property 'id' of null`

In `getAllParentWorkitems` the only references to `.id` are for the WI we are looking for the parent of, but we know that is good given the line `Looking for parents of WI [972]` in the log, so it must be the `if (!allRelatedWorkItems.find(element => element.id === id)) ` as this is the only other usage of `id` as a property. A potential parent element is a null, strange

All I can think to do is add an `is not empty check for  allRelatedWorkItems` but I can't see why this should only have shown up now.

As to running in the test harness, the task should behave identically, so you should be able to repo and experiment with dumping the payload and using a debugger.

We really need to see the payload to see what is going on here.

**Comment by nealeb4zod** [comment]
I'm not sure what the fault in dumpJsonPayload refers to?  Either way, the release notes are being generated, we're just getting a partial failure on the build task/build.

I'm still unable to reproduce in the test harness, it all completes as it should, however I've retrieved the payload from the pipeline along with the logs.  It is failing with a different work item this time - although it seems to be the last work item again.

[releasenoteslogs2.txt](https://github.com/rfennell/AzurePipelines/files/7728932/releasenoteslogs2.txt)
[payload (1).zip](https://github.com/rfennell/AzurePipelines/files/7728947/payload.1.zip)

**Comment by rfennell** [comment]
Sorry, I will try to explain again on the dumpJsonPayload issue. It seems your null id issue is also causing the optional payload logger to error when it tries to make a string representation of the JSON. But as you had logging off this had no effect on the task running to completion. This has shown that my logging dump logic could be better, also why am I trying to stringify the JSON when I don't use it? But that is a problem for me to sort out in a future release

As to your problem..... 

The log shows the issue is with the parent of 778

> 2021-12-16T16:21:45.6121976Z Looking for parents of WI [778]
> 2021-12-16T16:21:45.6126023Z The most common runtime reason for the task to fail is due API ECONNRESET issues. To avoid this failing the pipeline these will be treated as warnings and an attempt to generate any release notes possible
> 2021-12-16T16:21:45.6142476Z TypeError: Cannot read property 'id' of null

Looking in the dump I see this is 784 (ln126693), and as we would expect given the error there is no WI 784 in the dump

So why can't it load 784?

What do you see if you access the REST API in a brower with the following URL (edited with your org and project)

`https://<org>.visualstudio.com/<project>/_apis/wit/workitems/784`

I would expect a fully formed WI, but we might be able to spot something. If that does not work I will need to add extra logging into the find parents code. But it is strane yoy don't see the same issue with the test harness

**Comment by nealeb4zod** [comment]
Hi, thanks for you help so far, it's much appreciated.
Accessing the WI via the API returns a complete WI.  
The problem WI in the original log differs from the most recent one, could it be final WI that it processes that is the issue?
It could be an environmental thing, given that the test harness is A-OK, although it's a pretty standard build pipeline that it is in.

I've dumped the log from the test harness to a text file -  it looks very different to me from the pipeline log, so maybe there's a config setting that needs tweaked - I've tried to make them identical.
[testharnesslog.txt](https://github.com/rfennell/AzurePipelines/files/7735552/testharnesslog.txt)
